vowpalwabbit/automl.h

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+#pragma once
+
+#include "rand_state.h"
+#include "reductions_fwd.h"
+#include "distributionally_robust.h"
+#include "constant.h"  // NUM_NAMESPACES
+#include "metric_sink.h"
+#include "action_score.h"
+#include <map>
+#include <set>
+#include <queue>
+
+namespace VW
+{
+namespace automl
+{
+using namespace_index = unsigned char;
+using interaction_vec_t = std::vector<std::vector<namespace_index>>;
+
+VW::LEARNER::base_learner* automl_setup(VW::setup_base_i&);
+
+namespace details
+{
+void fail_if_enabled(vw&, const std::set<std::string>&);
+// void print_weights_nonzero(vw*, size_t, dense_parameters&);
+}  // namespace details
+
+constexpr size_t MAX_CONFIGS = 10;
+constexpr size_t CONGIGS_PER_CHAMP_CHANGE = 5;
+
+struct scored_config
+{
+  VW::distributionally_robust::ChiSquared chisq;
+  float ips = 0.0;
+  float last_w = 0.0;
+  float last_r = 0.0;
+  size_t update_count = 0;
+  size_t config_index = 0;
+  bool regular = false;
```

Comment:
regular as in non-champ?

Reply:
This is a newer idea which Paul suggested as an implementation of the `median` function in Algorithm 2 of Chacha. Basically we set half (or some number with the `--priority_challengers` flag) of the live configs as `regular`, and the other half as `priority`. The champ is a priority config, but is used for predictions while the other configs are not. When a priority config runs out of lease, its lease it doubled and it's able to keep running. When a regular config runs out of lease, we first compare it's `lower_bound` to each of the priority config's `ips`. If the lower bound is higher, we'll swap the priorities (priority becomes regular, regular becomes priority), then double its lease and keep running. If the `lower_bound` of the regular config can't beat the `ips` of any priority configs, then it is swapped out for some new config on the `index_queue`. The idea is that some live slots should be used for priority challengers to have more time to beat the champ, while others are used to swap out and try new configs more consistently.